src/providers/VoiceFocusProvider/index.tsx

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useState, useEffect, useMemo, useContext, createContext, useCallback } from 'react';
+import {
+  Device,
+  VoiceFocusTransformDevice,
+  VoiceFocusDeviceTransformer,
+  VoiceFocusSpec,
+  VoiceFocusDeviceOptions,
+} from 'amazon-chime-sdk-js';
+
+interface Props {
+  /** Determines how how you want Amazon Voice Focus to behave */
+  spec?: VoiceFocusSpec;
+  /** Configuration for additional named arguments, including `logger` and `preload` */
+  options?: VoiceFocusDeviceOptions;
+}
+
+interface VoiceFocusState {
+  supportsVoiceFocus: boolean;
+  addVoiceFocus: (device: Device) => Promise<Device | VoiceFocusTransformDevice>;
+}
+
+const VoiceFocusContext = createContext<VoiceFocusState | null>(null);
+
+const VoiceFocusProvider: React.FC<Props> = ({
+  spec,
+  options,
+  children
+}) => {
+  const [supportsVoiceFocus, setSupportsVoiceFocus] = useState(false);
+  const [voiceFocusDevice, setVoiceFocusDevice] = useState<VoiceFocusTransformDevice | null>(null);
+  const [voiceFocusTransformer, setVoiceFocusTransformer] = useState<VoiceFocusDeviceTransformer | null>(null);
+
+  useEffect(() => {
+    async function initVoiceFocus() {
+      try {
+        const isVoiceFocusSupported = await VoiceFocusDeviceTransformer.isSupported(spec, options);
+        if (isVoiceFocusSupported) {
+          const transformer = await getVoiceFocusDeviceTransformer();
+          if (transformer && transformer.isSupported()) {
+            setSupportsVoiceFocus(true);
+            return;
+          }
+        }
+      } catch (e) {
+        console.warn('Amazon Voice Focus is not supported.', e);
+      }
+      setSupportsVoiceFocus(false);
+    };
+    initVoiceFocus();
+  }, []);
+
+  useEffect(() => {
+    if (supportsVoiceFocus) {
+      console.log('Amazon Voice Focus is supported.');
+    } else {
+      console.warn('Amazon Voice Focus is not supported.');
+    }
+  }, [supportsVoiceFocus]);
+
+  async function getVoiceFocusDeviceTransformer(): Promise<VoiceFocusDeviceTransformer> {
+    if (voiceFocusTransformer) {
+      return voiceFocusTransformer;
+    } 
+
+    const transformer = await VoiceFocusDeviceTransformer.create(spec, options);
+    setVoiceFocusTransformer(transformer);
+    return transformer;
+  }
+
+  const addVoiceFocus = useCallback(async (device: Device): Promise<Device | VoiceFocusTransformDevice> => {
+    if (!voiceFocusDevice) {
+      return await createVoiceFocusDevice(device);
+    }
+    const vf = await voiceFocusDevice.chooseNewInnerDevice(device);
+    setVoiceFocusDevice(vf);
+    return vf;
+  }, [supportsVoiceFocus]);
```

Comment:
Should this dependency be `voiceFocusDevice` instead of `supportsVoiceFocus`?

Further, should we also check for the `supportsVoiceFocus` in this function and print a log if it is not supported and hence you cannot use VF or addVoidFocus? Thinking of a scenario if by mistake builders only use this provider and call `addVoiceFocus`.

Reply:
`voiceFocusDevice` dependency is not needed, since, we can create a new transform device based on the newer inner device and hence no dependency update is needed here.